From a748999e2015a9546f84d7169c6480ed8f6aa0ee Mon Sep 17 00:00:00 2001
From: Philippe Nunes <philippe.nunes@linux.intel.com>
Date: Tue, 18 Sep 2012 17:35:02 +0200
Subject: [PATCH 01/23] desc: CP name is imc-pr3 for Intel device
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/desc.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/desc.c b/src/desc.c
index d165ca0..7a5ef93 100644
--- a/src/desc.c
+++ b/src/desc.c
@@ -123,6 +123,18 @@ static int _get_cp_name(char **name)
 	uname(&u);
 
 	dbg("u.nodename : [ %s ]", u.nodename);
+	dbg("u.__domainname : [ %s ]", u.__domainname);
+	dbg("u.machine : [ %s ]", u.machine);
+	dbg("u.release : [ %s ]", u.release);
+	dbg("u.sysname : [ %s ]", u.sysname);S
+	dbg("u.version : [ %s ]", u.version);
+
+	/* By default, SUNRISE IMC modem is returned */
+	if (!strcmp(u.nodename, "(none)")) {
+		*name = g_new0(char, 8);
+		strcpy(*name, "imc-pr3");
+		return 8;
+	}
 
 	for (i = 0; svnet1_models[i]; i++) {
 		if (!strcmp(u.nodename, svnet1_models[i])) {
-- 
1.7.10.4

