From 389f7a04eefccdec5f2597df0d6c836a0010b268 Mon Sep 17 00:00:00 2001
From: Philippe Nunes <philippe.nunes@linux.intel.com>
Date: Tue, 18 Sep 2012 17:47:49 +0200
Subject: [PATCH 02/23] s_modem: CGMR response parsing is compatible with IMC
 PR3 modem
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/s_modem.c |   21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/src/s_modem.c b/src/s_modem.c
index 88a06a7..edd1ca9 100644
--- a/src/s_modem.c
+++ b/src/s_modem.c
@@ -360,20 +360,23 @@ static void on_response_version(TcorePending *p, int data_len, const void *data,
 		if (resp->lines) {
 			line = (const char *) resp->lines->data;
 			tokens = tcore_at_tok_new(line);
-			if (g_slist_length(tokens) != 5) {
+			if (g_slist_length(tokens) == 1) {
+				swver = g_slist_nth_data(tokens, 0);
+				dbg("version: sw=[%s]", swver);
+			} else if (g_slist_length(tokens) == 5) {
+				swver = g_slist_nth_data(tokens, 0);
+				hwver = g_slist_nth_data(tokens, 1);
+				caldate = g_slist_nth_data(tokens, 2);
+				pcode = g_slist_nth_data(tokens, 3);
+				id = g_slist_nth_data(tokens, 4);
+
+				dbg("version: sw=[%s], hw=[%s], rf_cal=[%s], product_code=[%s], model_id=[%s]", swver, hwver, caldate, pcode, id);
+			} else {
 				msg("invalid message");
 				goto OUT;
 			}
 		}
 
-		swver = g_slist_nth_data(tokens, 0);
-		hwver = g_slist_nth_data(tokens, 1);
-		caldate = g_slist_nth_data(tokens, 2);
-		pcode = g_slist_nth_data(tokens, 3);
-		id = g_slist_nth_data(tokens, 4);
-
-		dbg("version: sw=[%s], hw=[%s], rf_cal=[%s], product_code=[%s], model_id=[%s]", swver, hwver, caldate, pcode, id);
-
 		vi = calloc(sizeof(TelMiscVersionInformation), 1);
 		if (NULL != swver)
 			memcpy(vi->szSwVersion, swver, strlen(swver));
-- 
1.7.10.4

