From 7e83caa59a5a98f169a45b58b9767fc8f65e61bf Mon Sep 17 00:00:00 2001
From: Philippe Nunes <philippe.nunes@linux.intel.com>
Date: Tue, 18 Sep 2012 17:45:58 +0200
Subject: [PATCH 03/23] s-modem: Cleanup
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Conflicts:
	src/s_modem.c
---
 src/s_modem.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/s_modem.c b/src/s_modem.c
index edd1ca9..dd8899b 100644
--- a/src/s_modem.c
+++ b/src/s_modem.c
@@ -250,6 +250,7 @@ static void on_response_set_flight_mode(TcorePending *p, int data_len, const voi
 		}
 	} else {
 		dbg("Sending response for Flight mode operation");
+		tcore_user_request_send_response(ur, TRESP_MODEM_SET_FLIGHTMODE, sizeof(struct tresp_modem_set_flightmode), &res);
 
 		req_data = tcore_user_request_ref_data(ur, NULL);
 
@@ -787,7 +788,7 @@ static TReturn get_imei(CoreObject *o, UserRequest *ur)
 	TcorePending *pending = NULL;
 
 	hal = tcore_object_get_hal(o);
-	if(FALSE == tcore_hal_get_power_state(hal)){
+	if (FALSE == tcore_hal_get_power_state(hal)) {
 		dbg("cp not ready/n");
 		return TCORE_RETURN_ENOSYS;
 	}
-- 
1.7.10.4

