From 18e1f3c2d0e927ff7ff4d8e6c6dd16ba701e9523 Mon Sep 17 00:00:00 2001
From: Philippe Nunes <philippe.nunes@linux.intel.com>
Date: Tue, 18 Sep 2012 17:50:08 +0200
Subject: [PATCH 04/23] s-modem: Add notification hook for SIM status
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/s_modem.c |   16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/src/s_modem.c b/src/s_modem.c
index dd8899b..2c71457 100644
--- a/src/s_modem.c
+++ b/src/s_modem.c
@@ -465,6 +465,20 @@ OUT:
 }
 
 
+static enum tcore_hook_return on_hook_sim_status(Server *s, CoreObject *source, enum tcore_notification_command command,
+											   unsigned int data_len, void *data, void *user_data)
+{
+	const struct tnoti_sim_status *sim = data;
+
+	if (sim->sim_status == SIM_STATUS_INIT_COMPLETED) {
+		dbg("SIM ready for attach");
+		dbg("Enable STK and Fetching of proactive Commands");
+		prepare_and_send_pending_request(tcore_object_ref_plugin(source), "sat", "AT+CFUN=6", NULL, TCORE_AT_NO_RESULT, on_response_enable_proactive_command);
+		prepare_and_send_pending_request(tcore_object_ref_plugin(source), "umts_network", "AT+COPS=0", NULL, TCORE_AT_NO_RESULT, on_response_bootup_subscription);
+	}
+
+	return TCORE_HOOK_RETURN_CONTINUE;
+}
 
 gboolean on_event_modem_power(TcoreAT *at, const char *line, TcorePlugin *p)
 {
@@ -917,6 +931,8 @@ gboolean s_modem_init(TcorePlugin *p, TcoreHal *h)
 	dbg("Registering for +XSIM event");
 	tcore_object_add_callback(o, "+XSIM", on_event_bootup_sim_status, NULL);
 
+	tcore_server_add_notification_hook(tcore_plugin_ref_server(p), TNOTI_SIM_STATUS, on_hook_sim_status, o);
+
 	return TRUE;
 }
 
-- 
1.7.10.4

