From 86510097b43f9a60f50547511fb176cf6d97e9b2 Mon Sep 17 00:00:00 2001
From: Philippe Nunes <philippe.nunes@linux.intel.com>
Date: Thu, 20 Sep 2012 17:28:07 +0200
Subject: [PATCH 06/23] common: Fix warning errors
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/s_common.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/s_common.c b/src/s_common.c
index 0c83ac5..9b8c4a3 100644
--- a/src/s_common.c
+++ b/src/s_common.c
@@ -226,9 +226,9 @@ char _util_unpackb(const char *src, int pos, int len)
 	rshift = MAX(8 - (pos + len), 0);
 
 	if (rshift > 0) {
-		result = MASK_AND_SHIFT(len, pos, rshift, *src);
+		result = MASK_AND_SHIFT(len, pos, rshift, (unsigned char)*src);
 	} else {
-		result = MASK(8 - pos, pos, *src);
+		result = MASK(8 - pos, pos, (unsigned char)*src);
 		src++;
 		len -= 8 - pos;
 
-- 
1.7.10.4

