From a2b5d447eb38825e0af9e57dde07b6e941d48389 Mon Sep 17 00:00:00 2001
From: Philippe Nunes <philippe.nunes@linux.intel.com>
Date: Wed, 26 Sep 2012 17:26:42 +0200
Subject: [PATCH 09/23] s_modem: Add XGENDATA query to get firmware
 information
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/s_modem.c |   30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/src/s_modem.c b/src/s_modem.c
index 2c71457..22adfa3 100644
--- a/src/s_modem.c
+++ b/src/s_modem.c
@@ -275,6 +275,33 @@ static void on_response_set_flight_mode(TcorePending *p, int data_len, const voi
 	tcore_at_tok_free(tokens);
 }
 
+static void on_response_xgendata(TcorePending *p, int data_len, const void *data, void *user_data)
+{
+	const TcoreATResponse *resp = data;
+	GSList *tokens = NULL;
+	const char *line;
+
+	if (resp->success > 0) {
+		dbg("RESPONSE OK");
+		if (resp->lines) {
+			line = (const char *) resp->lines->data;
+			tokens = tcore_at_tok_new(line);
+			if (g_slist_length(tokens) != 1) {
+				msg("invalid message");
+				goto OUT;
+			}
+		}
+
+		dbg("XGENDATA = [%s]", g_slist_nth_data(tokens, 0));
+	}
+
+OUT:
+	if (tokens != NULL)
+		tcore_at_tok_free(tokens);
+
+	return;
+}
+
 static void on_response_imei(TcorePending *p, int data_len, const void *data, void *user_data)
 {
 	const TcoreATResponse *resp = data;
@@ -511,6 +538,9 @@ gboolean on_event_modem_power(TcoreAT *at, const char *line, TcorePlugin *p)
 	/* Get Version Number  */
 	prepare_and_send_pending_request(p, "modem", "AT+CGMR", NULL, TCORE_AT_SINGLELINE, on_response_version);
 
+	/* Get XGENDATA info */
+	prepare_and_send_pending_request(p, "modem", "AT+XGENDATA", "+XGENDATA:", TCORE_AT_MULTILINE, on_response_xgendata);
+
 	tcore_modem_set_powered(o, TRUE);
 
 	modem_power.state = MODEM_STATE_ONLINE;
-- 
1.7.10.4

