From 0ab2df1acdc37d1ada019d0e4c1ccb81913637ef Mon Sep 17 00:00:00 2001
From: Philippe Nunes <philippe.nunes@linux.intel.com>
Date: Wed, 26 Sep 2012 17:30:40 +0200
Subject: [PATCH 10/23] s_sim: Extend XSIMSTATE parsing to get SMS service
 state
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/s_sim.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/src/s_sim.c b/src/s_sim.c
index 1d7e5c6..7bff6dd 100644
--- a/src/s_sim.c
+++ b/src/s_sim.c
@@ -28,12 +28,14 @@
 #include <plugin.h>
 #include <queue.h>
 #include <co_sim.h>
+#include <co_sms.h>
 #include <storage.h>
 #include <user_request.h>
 #include <server.h>
 #include <at.h>
 
 #include "s_common.h"
+#include "s_sms.h"
 #include "s_sim.h"
 
 #define ID_RESERVED_AT 0x0229
@@ -1932,6 +1934,7 @@ static gboolean on_event_pin_status(CoreObject *o, const void *event_info, void
 	GSList *lines = NULL;
 	const char *line = NULL;
 	int sim_state = 0;
+	int sms_state = 0;
 
 	dbg(" Function entry ");
 
@@ -1948,6 +1951,7 @@ static gboolean on_event_pin_status(CoreObject *o, const void *event_info, void
 
 	if (g_slist_length(tokens) == 4) {
 		sim_state = atoi(g_slist_nth_data(tokens, 1));
+		sms_state = atoi(g_slist_nth_data(tokens, 3));
 	} else if (g_slist_length(tokens) == 1)
 		sim_state = atoi(g_slist_nth_data(tokens, 0));
 	else {
@@ -2023,6 +2027,20 @@ static gboolean on_event_pin_status(CoreObject *o, const void *event_info, void
 			_get_sim_type(o);
 		else
 			_sim_status_update(o, sim_status);
+
+		if (sms_state) {
+			struct tnoti_sms_ready_status readyStatusInfo = {0, };
+			CoreObject *sms;
+			TcorePlugin *plugin;
+
+			dbg("SMS Ready");
+			readyStatusInfo.status = TRUE;
+			plugin = tcore_object_ref_plugin(o);
+			sms = tcore_plugin_ref_core_object(plugin, "umts_sms");
+			tcore_sms_set_ready_status(sms, readyStatusInfo.status);
+
+			tcore_server_send_notification(tcore_plugin_ref_server(plugin), sms, TNOTI_SMS_DEVICE_READY, sizeof(struct tnoti_sms_ready_status), &readyStatusInfo);
+		}
 		break;
 
 	case SIM_STATUS_INITIALIZING:
-- 
1.7.10.4

