From ee0731a4b1c6cb4727910320d9b8e5f96bad9f4b Mon Sep 17 00:00:00 2001
From: Caiwen Zhang <caiwen.zhang@intel.com>
Date: Fri, 28 Sep 2012 18:28:10 +0800
Subject: [PATCH 11/23] set modem power saving mode
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/s_modem.c |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/s_modem.c b/src/s_modem.c
index 22adfa3..a0fe4dc 100644
--- a/src/s_modem.c
+++ b/src/s_modem.c
@@ -703,6 +703,8 @@ static void on_response_enable_logging(TcorePending *p, int data_len, const void
 	dbg("Calling setup_mux");
 	setup_mux(tcore_pending_ref_core_object(p));
 
+	/* sets the powersaving-mode */
+	prepare_and_send_pending_request(plugin, "modem", "at+xpow=0,0,0", NULL, TCORE_AT_NO_RESULT, NULL);
 
 	dbg("Exit");
 	return;
-- 
1.7.10.4

