From e8ae3b0057dad62602dd539c7e8ea5839d527e8a Mon Sep 17 00:00:00 2001
From: Nicolas Bertrand <nicolas.bertrand@linux.intel.com>
Date: Fri, 8 Feb 2013 11:21:45 +0100
Subject: [PATCH] Use plugin mfld-blackbay
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

---
 src/desc.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/desc.c b/src/desc.c
index 61833f9..01385a2 100644
--- a/src/desc.c
+++ b/src/desc.c
@@ -86,8 +86,8 @@ static int _get_cp_name(char **name)
 
 	/* By default, SUNRISE IMC modem is returned */
 	if (!strcmp(u.nodename, "(none)")) {
-		*name = g_new0(char, 8);
-		strcpy(*name, "imc-pr3");
+		*name = g_new0(char, 14);
+		strcpy(*name, "mfld-blackbay");
 		return 8;
 	}
 
-- 
1.7.10.4

